File: ios/CodeMatch/Services/FeedbackPlayer.swift
import AVFoundation
import UIKit

/// 効果音設定。UserDefaultsで永続化し、設定画面とFeedbackPlayerで共有する。
enum FeedbackSettings {
    static let volumeKey = "feedbackVolume"
    static let successSoundKey = "successSound"
    static let failureSoundKey = "failureSound"
    static let defaultVolume = 1.0

    static var volume: Double {
        UserDefaults.standard.object(forKey: volumeKey) as? Double ?? defaultVolume
    }

    static var successSound: SuccessSound {
        SuccessSound(rawValue: UserDefaults.standard.string(forKey: successSoundKey) ?? "") ?? .posBeep
    }

    static var failureSound: FailureSound {
        FailureSound(rawValue: UserDefaults.standard.string(forKey: failureSoundKey) ?? "") ?? .alarm
    }
}

enum SuccessSound: String, CaseIterable, Identifiable {
    case sample1
    case sample2
    case posBeep
    case doubleBeep
    case chime

    var id: String { rawValue }

    var label: String {
        switch self {
        case .sample1: AppLocalization.string("サウンド1")
        case .sample2: AppLocalization.string("サウンド2")
        case .posBeep: AppLocalization.string("ピッ（POSレジ風・標準）")
        case .doubleBeep: AppLocalization.string("ピピッ（2回）")
        case .chime: AppLocalization.string("チャイム（3音）")
        }
    }

    /// バンドル内の音源ファイル名。nilなら合成音を使う。
    var fileName: String? {
        switch self {
        case .sample1: "success1"
        case .sample2: "success2"
        case .posBeep, .doubleBeep, .chime: nil
        }
    }
}

enum FailureSound: String, CaseIterable, Identifiable {
    case failSample
    case buzzer
    case alarm
    case descend

    var id: String { rawValue }

    var label: String {
        switch self {
        case .failSample: AppLocalization.string("サウンド1")
        case .buzzer: AppLocalization.string("ブブー（ブザー）")
        case .alarm: AppLocalization.string("ピピピピ（アラーム・標準）")
        case .descend: AppLocalization.string("ブーー（下降音）")
        }
    }

    /// バンドル内の音源ファイル名。nilなら合成音を使う。
    var fileName: String? {
        switch self {
        case .failSample: "Fail1"
        case .buzzer, .alarm, .descend: nil
        }
    }
}

@MainActor
final class FeedbackPlayer {
    private struct Tone {
        let frequency: Double
        let duration: TimeInterval
        let amplitude: Float
        /// trueのとき奇数倍音を加えてPOSレジのような鋭く通る音にする
        var piercing: Bool = false
    }

    private let audioEngine = AVAudioEngine()
    private let audioPlayer = AVAudioPlayerNode()
    private let sampleRate = 44_100.0
    private lazy var audioFormat = AVAudioFormat(
        standardFormatWithSampleRate: sampleRate,
        channels: 1
    )!
    /// バンドル音源のプレイヤーキャッシュ(ファイル名 → プレイヤー)
    private var filePlayers: [String: AVAudioPlayer] = [:]

    /// AVAudioEngineのセットアップは重く、生成のたびにオーディオ通知を発生させて
    /// SwiftUIの再描画ループを誘発しうるため、アプリ全体で1インスタンスを共有する。
    static let shared = FeedbackPlayer()

    private init() {
        audioEngine.attach(audioPlayer)
        audioEngine.connect(audioPlayer, to: audioEngine.mainMixerNode, format: audioFormat)
        audioEngine.prepare()
    }

    func scanAccepted() {
        UIImpactFeedbackGenerator(style: .medium).impactOccurred()
        // 各コード受理時は控えめな短いブリップ(最終判定音より小さめ)
        play([
            Tone(frequency: 1_567.98, duration: 0.06, amplitude: 0.45, piercing: true)
        ], gap: 0, volumeScale: 0.6)
    }

    /// 入力順序または業務フォーマットが違う読取を、最終照合の不一致音とは区別して通知する。
    func invalidScan() {
        UINotificationFeedbackGenerator().notificationOccurred(.warning)
        play([
            Tone(frequency: 330, duration: 0.09, amplitude: 0.65),
            Tone(frequency: 330, duration: 0.09, amplitude: 0.65)
        ], gap: 0.06, volumeScale: 0.7)
    }

    func success(after delay: TimeInterval = 0) {
        UINotificationFeedbackGenerator().notificationOccurred(.success)

        let playSound: () -> Void = { [weak self] in
            guard let self else { return }
            let sound = FeedbackSettings.successSound
            if let fileName = sound.fileName {
                self.playFile(named: fileName)
                return
            }
            switch sound {
            case .sample1, .sample2:
                break
            case .posBeep:
                self.play([
                    Tone(frequency: 2_600, duration: 0.12, amplitude: 0.95, piercing: true)
                ], gap: 0)
            case .doubleBeep:
                self.play([
                    Tone(frequency: 2_600, duration: 0.08, amplitude: 0.95, piercing: true),
                    Tone(frequency: 2_600, duration: 0.08, amplitude: 0.95, piercing: true)
                ], gap: 0.06)
            case .chime:
                self.play([
                    Tone(frequency: 523.25, duration: 0.09, amplitude: 0.80),
                    Tone(frequency: 659.25, duration: 0.09, amplitude: 0.85),
                    Tone(frequency: 783.99, duration: 0.18, amplitude: 0.90)
                ], gap: 0.035)
            }
        }

        if delay > 0 {
            DispatchQueue.main.asyncAfter(deadline: .now() + delay, execute: playSound)
        } else {
            playSound()
        }
    }

    func failure() {
        UINotificationFeedbackGenerator().notificationOccurred(.error)
        let sound = FeedbackSettings.failureSound
        if let fileName = sound.fileName {
            playFile(named: fileName)
            return
        }
        switch sound {
        case .failSample:
            break
        case .buzzer:
            play([
                Tone(frequency: 165, duration: 0.16, amplitude: 0.95, piercing: true),
                Tone(frequency: 165, duration: 0.42, amplitude: 0.95, piercing: true)
            ], gap: 0.07)
        case .alarm:
            play([
                Tone(frequency: 980, duration: 0.11, amplitude: 0.92, piercing: true),
                Tone(frequency: 980, duration: 0.11, amplitude: 0.92, piercing: true),
                Tone(frequency: 980, duration: 0.11, amplitude: 0.92, piercing: true),
                Tone(frequency: 980, duration: 0.11, amplitude: 0.92, piercing: true)
            ], gap: 0.09)
        case .descend:
            play([
                Tone(frequency: 440, duration: 0.18, amplitude: 0.92, piercing: true),
                Tone(frequency: 220, duration: 0.45, amplitude: 0.95, piercing: true)
            ], gap: 0.04)
        }
    }

    /// 設定画面のプレビュー再生用
    func preview(success sound: SuccessSound) {
        let saved = FeedbackSettings.successSound
        UserDefaults.standard.set(sound.rawValue, forKey: FeedbackSettings.successSoundKey)
        success()
        if saved != sound {
            UserDefaults.standard.set(saved.rawValue, forKey: FeedbackSettings.successSoundKey)
        }
    }

    func preview(failure sound: FailureSound) {
        let saved = FeedbackSettings.failureSound
        UserDefaults.standard.set(sound.rawValue, forKey: FeedbackSettings.failureSoundKey)
        failure()
        if saved != sound {
            UserDefaults.standard.set(saved.rawValue, forKey: FeedbackSettings.failureSoundKey)
        }
    }

    /// バンドル内の音源ファイルを設定音量で再生する
    private func playFile(named name: String) {
        guard let url = Bundle.main.url(forResource: name, withExtension: "mp3") else { return }
        do {
            let session = AVAudioSession.sharedInstance()
            // 工場現場向け: マナーモード(サイレントスイッチ)でも判定音を再生する
            try session.setCategory(.playback, mode: .default, options: [.mixWithOthers])
            try session.setActive(true)

            let player: AVAudioPlayer
            if let cached = filePlayers[name] {
                player = cached
            } else {
                player = try AVAudioPlayer(contentsOf: url)
                player.prepareToPlay()
                filePlayers[name] = player
            }
            player.volume = Float(FeedbackSettings.volume)
            player.currentTime = 0
            player.play()
        } catch {
            // Haptic feedback still communicates the result if audio is unavailable.
        }
    }

    private func play(_ tones: [Tone], gap: TimeInterval, volumeScale: Float = 1.0) {
        guard let buffer = makeBuffer(tones: tones, gap: gap) else { return }

        do {
            let session = AVAudioSession.sharedInstance()
            // 工場現場向け: マナーモード(サイレントスイッチ)でも判定音を再生する
            try session.setCategory(.playback, mode: .default, options: [.mixWithOthers])
            try session.setActive(true)
            if !audioEngine.isRunning {
                try audioEngine.start()
            }

            audioPlayer.volume = Float(FeedbackSettings.volume) * volumeScale
            audioPlayer.stop()
            audioPlayer.scheduleBuffer(buffer, at: nil, options: .interrupts)
            audioPlayer.play()
        } catch {
            // Haptic feedback still communicates the result if audio is unavailable.
        }
    }

    private func makeBuffer(tones: [Tone], gap: TimeInterval) -> AVAudioPCMBuffer? {
        let gapFrames = Int(gap * sampleRate)
        let toneFrames = tones.map { Int($0.duration * sampleRate) }
        let totalFrames = toneFrames.reduce(0, +) + (max(tones.count - 1, 0) * gapFrames)
        guard totalFrames > 0,
              let buffer = AVAudioPCMBuffer(
                pcmFormat: audioFormat,
                frameCapacity: AVAudioFrameCount(totalFrames)
              ),
              let samples = buffer.floatChannelData?[0]
        else { return nil }

        buffer.frameLength = AVAudioFrameCount(totalFrames)
        var frameOffset = 0

        for (index, tone) in tones.enumerated() {
            let frameCount = toneFrames[index]
            let edgeFrames = min(Int(0.006 * sampleRate), max(frameCount / 2, 1))

            for frame in 0..<frameCount {
                let attack = min(Float(frame) / Float(edgeFrames), 1)
                let release = min(Float(frameCount - frame - 1) / Float(edgeFrames), 1)
                let envelope = max(min(attack, release), 0)
                let phase = 2 * Double.pi * tone.frequency * Double(frame) / sampleRate

                var sample = sin(phase)
                if tone.piercing {
                    // 奇数倍音を加えた擬似矩形波。小さなスピーカーでも通る音になる
                    sample += sin(phase * 3) / 3 + sin(phase * 5) / 5
                    sample *= 0.85
                }
                samples[frameOffset + frame] = Float(sample) * tone.amplitude * envelope
            }

            frameOffset += frameCount
            if index < tones.count - 1 {
                for frame in 0..<gapFrames {
                    samples[frameOffset + frame] = 0
                }
                frameOffset += gapFrames
            }
        }

        return buffer
    }
}

File: android/app/src/main/java/jp/rimtty/codematch/feedback/FeedbackPlayer.kt
package jp.rimtty.codematch.feedback

import android.content.Context
import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import android.media.MediaPlayer
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.provider.Settings
import dagger.hilt.android.qualifiers.ApplicationContext
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong
import javax.inject.Inject
import javax.inject.Singleton
import jp.rimtty.codematch.R
import jp.rimtty.codematch.core.model.FailureSound
import jp.rimtty.codematch.core.model.SuccessSound
import kotlin.math.PI
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sin

/** The haptic cue paired with a user-visible feedback state. */
enum class FeedbackHaptic {
    MEDIUM,
    WARNING,
    SUCCESS,
    ERROR,
}

/** One synthesized tone, expressed independently of Android audio classes. */
data class FeedbackTone(
    val frequencyHz: Double,
    val durationMillis: Int,
    val amplitude: Float,
    val piercing: Boolean = false,
)

/**
 * Platform-neutral feedback contract shared by the player and JVM tests.
 *
 * Asset names are Android raw-resource names without an extension. A cue with
 * an asset has no synthesized tones; a synthesized cue has one or more tones.
 */
data class FeedbackCue(
    val tones: List<FeedbackTone>,
    val gapMillis: Int,
    val volumeScale: Float,
    val haptic: FeedbackHaptic,
    val assetName: String? = null,
) {
    init {
        require(tones.isNotEmpty() xor (assetName != null)) {
            "A feedback cue must use either tones or an audio asset"
        }
        require(gapMillis >= 0) { "gapMillis must not be negative" }
        require(volumeScale in 0f..1f) { "volumeScale must be between 0 and 1" }
    }

    /** Audio is intentionally suppressed at zero volume; haptic is not. */
    fun shouldPlayAudio(volume: Float): Boolean = volume > 0f &&
        (assetName != null || tones.isNotEmpty())
}

/**
 * iOS FeedbackPlayer.swift's frequency, timing, asset, and haptic contract.
 * Keep this object free of Android framework types so its values are
 * deterministic in local unit tests.
 */
object FeedbackContract {
    val scanAccepted = FeedbackCue(
        tones = listOf(
            FeedbackTone(
                frequencyHz = 1_567.98,
                durationMillis = 60,
                amplitude = 0.45f,
                piercing = true,
            ),
        ),
        gapMillis = 0,
        volumeScale = 0.6f,
        haptic = FeedbackHaptic.MEDIUM,
    )

    val invalidScan = FeedbackCue(
        tones = listOf(
            FeedbackTone(330.0, 90, 0.65f),
            FeedbackTone(330.0, 90, 0.65f),
        ),
        gapMillis = 60,
        volumeScale = 0.7f,
        haptic = FeedbackHaptic.WARNING,
    )

    fun success(sound: SuccessSound): FeedbackCue = when (sound) {
        SuccessSound.SAMPLE_1 -> asset("success1")
        SuccessSound.SAMPLE_2 -> asset("success2")
        SuccessSound.POS_BEEP -> synthesized(
            tones = listOf(FeedbackTone(2_600.0, 120, 0.95f, piercing = true)),
        )
        SuccessSound.DOUBLE_BEEP -> synthesized(
            tones = listOf(
                FeedbackTone(2_600.0, 80, 0.95f, piercing = true),
                FeedbackTone(2_600.0, 80, 0.95f, piercing = true),
            ),
            gapMillis = 60,
        )
        SuccessSound.CHIME -> synthesized(
            tones = listOf(
                FeedbackTone(523.25, 90, 0.80f),
                FeedbackTone(659.25, 90, 0.85f),
                FeedbackTone(783.99, 180, 0.90f),
            ),
            gapMillis = 35,
        )
    }

    fun failure(sound: FailureSound): FeedbackCue = when (sound) {
        FailureSound.FAIL_SAMPLE -> asset("fail1", FeedbackHaptic.ERROR)
        FailureSound.BUZZER -> synthesized(
            tones = listOf(
                FeedbackTone(165.0, 160, 0.95f, piercing = true),
                FeedbackTone(165.0, 420, 0.95f, piercing = true),
            ),
            gapMillis = 70,
            haptic = FeedbackHaptic.ERROR,
        )
        FailureSound.ALARM -> synthesized(
            tones = List(4) { FeedbackTone(980.0, 110, 0.92f, piercing = true) },
            gapMillis = 90,
            haptic = FeedbackHaptic.ERROR,
        )
        FailureSound.DESCEND -> synthesized(
            tones = listOf(
                FeedbackTone(440.0, 180, 0.92f, piercing = true),
                FeedbackTone(220.0, 450, 0.95f, piercing = true),
            ),
            gapMillis = 40,
            haptic = FeedbackHaptic.ERROR,
        )
    }

    private fun asset(
        name: String,
        haptic: FeedbackHaptic = FeedbackHaptic.SUCCESS,
    ): FeedbackCue = FeedbackCue(
        tones = emptyList(),
        gapMillis = 0,
        volumeScale = 1.0f,
        haptic = haptic,
        assetName = name,
    )

    private fun synthesized(
        tones: List<FeedbackTone>,
        gapMillis: Int = 0,
        haptic: FeedbackHaptic = FeedbackHaptic.SUCCESS,
    ): FeedbackCue = FeedbackCue(
        tones = tones,
        gapMillis = gapMillis,
        volumeScale = 1.0f,
        haptic = haptic,
    )
}

/**
 * Plays the feedback contract without doing audio work on the main thread.
 *
 * A new cue interrupts an earlier cue, matching the iOS audio player's
 * `.interrupts` scheduling behavior. Haptic dispatch is independent of volume
 * so a 0% volume setting still communicates all four states by vibration.
 */
@Singleton
class FeedbackPlayer @Inject constructor(
    @param:ApplicationContext private val context: Context,
) {
    private class MediaPlayback(val player: MediaPlayer) {
        val released = AtomicBoolean(false)

        @Volatile
        var deadline: ScheduledFuture<*>? = null
    }

    private val audioExecutor: ExecutorService = Executors.newSingleThreadExecutor { runnable ->
        Thread(runnable, "codematch-feedback-audio").apply { isDaemon = true }
    }
    private val mediaDeadlineExecutor: ScheduledExecutorService =
        Executors.newSingleThreadScheduledExecutor { runnable ->
            Thread(runnable, "codematch-feedback-deadline").apply { isDaemon = true }
        }
    private val generation = AtomicLong(0)

    @Volatile
    private var currentTrack: AudioTrack? = null

    @Volatile
    private var currentMediaPlayback: MediaPlayback? = null

    fun playScanAccepted(volume: Float) {
        play(FeedbackContract.scanAccepted, volume)
    }

    fun playInvalidScan(volume: Float = 1.0f) {
        play(FeedbackContract.invalidScan, volume)
    }

    fun playSuccess(
        sound: SuccessSound,
        volume: Float,
        includeHaptic: Boolean = true,
    ) {
        play(FeedbackContract.success(sound), volume, includeHaptic)
    }

    fun playFailure(
        sound: FailureSound,
        volume: Float,
        includeHaptic: Boolean = true,
    ) {
        play(FeedbackContract.failure(sound), volume, includeHaptic)
    }

    private fun play(cue: FeedbackCue, volume: Float, includeHaptic: Boolean = true) {
        if (includeHaptic) playHaptic(cue.haptic)

        val token = generation.incrementAndGet()
        // Keep all MediaPlayer/AudioTrack operations off the caller (normally
        // the main/UI) thread. The ordered stop task still interrupts the
        // previous cue before this cue starts on the single audio executor.
        audioExecutor.execute {
            if (token != generation.get()) return@execute
            stopCurrentAudio()
            if (!cue.shouldPlayAudio(volume)) return@execute

            if (cue.assetName != null) {
                playAsset(cue.assetName, volume.coerceIn(0f, 1f), token)
            } else {
                playSynthesized(cue, volume.coerceIn(0f, 1f), token)
            }
        }
    }

    private fun playAsset(name: String, volume: Float, token: Long) {
        val resourceId = when (name) {
            "success1" -> R.raw.success1
            "success2" -> R.raw.success2
            "fail1" -> R.raw.fail1
            else -> return
        }
        val player = runCatching { MediaPlayer.create(context, resourceId) }.getOrNull() ?: return
        val playback = MediaPlayback(player)
        if (token != generation.get()) {
            releaseMediaPlayback(playback)
            return
        }
        val listenersInstalled = runCatching {
            player.setVolume(volume, volume)
            player.setOnCompletionListener {
                releaseMediaPlaybackAsync(playback)
            }
            player.setOnErrorListener { _, _, _ ->
                releaseMediaPlaybackAsync(playback)
                true
            }
        }.isSuccess
        if (!listenersInstalled) {
            releaseMediaPlayback(playback)
            return
        }
        val shouldRelease = synchronized(this) {
            if (token != generation.get()) {
                true
            } else {
                currentMediaPlayback = playback
                false
            }
        }
        if (shouldRelease) {
            releaseMediaPlayback(playback)
            return
        }

        val durationMillis = runCatching { player.duration.toLong() }
            .getOrDefault(0L)
            .coerceAtLeast(0L)
        playback.deadline = runCatching {
            mediaDeadlineExecutor.schedule(
                { releaseMediaPlayback(playback) },
                durationMillis + AUDIO_COMPLETION_MARGIN_MILLIS,
                TimeUnit.MILLISECONDS,
            )
        }.getOrNull()
        if (playback.deadline == null) {
            releaseMediaPlayback(playback)
            return
        }
        if (token != generation.get()) {
            releaseMediaPlayback(playback)
            return
        }

        runCatching { player.start() }
            .onFailure { releaseMediaPlayback(playback) }
    }

    /** MediaPlayer callbacks may be delivered on the main looper. */
    private fun releaseMediaPlaybackAsync(playback: MediaPlayback) {
        runCatching {
            audioExecutor.execute { releaseMediaPlayback(playback) }
        }.onFailure {
            // The deadline/new-cue paths remain responsible for cleanup if
            // the executor is no longer accepting callback work.
        }
    }

    private fun playSynthesized(cue: FeedbackCue, volume: Float, token: Long) {
        val samples = buildSamples(cue, volume)
        if (samples.isEmpty() || token != generation.get()) return

        val minBufferSize = AudioTrack.getMinBufferSize(
            SAMPLE_RATE,
            AudioFormat.CHANNEL_OUT_MONO,
            AudioFormat.ENCODING_PCM_16BIT,
        ).coerceAtLeast(1)
        val track = runCatching {
            AudioTrack.Builder()
                .setAudioAttributes(
                    AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_ASSISTANCE_SONIFICATION)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                        .build(),
                )
                .setAudioFormat(
                    AudioFormat.Builder()
                        .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                        .setSampleRate(SAMPLE_RATE)
                        .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                        .build(),
                )
                .setBufferSizeInBytes(max(samples.size * 2, minBufferSize))
                .setTransferMode(AudioTrack.MODE_STATIC)
                .build()
        }.getOrNull() ?: return

        if (token != generation.get()) {
            runCatching { track.release() }
            return
        }
        val written = runCatching { track.write(samples, 0, samples.size) }.getOrDefault(-1)
        if (written != samples.size || token != generation.get()) {
            runCatching { track.release() }
            return
        }
        synchronized(this) {
            if (token != generation.get()) {
                runCatching { track.release() }
                return
            }
            currentTrack = track
        }
        try {
            track.play()
            val durationMillis = cue.tones.sumOf { it.durationMillis } +
                cue.gapMillis * max(cue.tones.size - 1, 0)
            val deadlineNanos = System.nanoTime() +
                (durationMillis + AUDIO_COMPLETION_MARGIN_MILLIS) * 1_000_000L
            // Some devices keep MODE_STATIC in PLAYSTATE_PLAYING after the
            // final frame. Use both playback head and a hard deadline so every
            // cue is released even when that state is sticky.
            while (
                token == generation.get() &&
                track.playbackHeadPosition < samples.size &&
                System.nanoTime() < deadlineNanos
            ) {
                Thread.sleep(AUDIO_POLL_MILLIS)
            }
        } catch (_: InterruptedException) {
            Thread.currentThread().interrupt()
        } catch (_: RuntimeException) {
            // A transient audio-route failure must not affect scan state.
        } finally {
            synchronized(this) {
                if (currentTrack === track) currentTrack = null
            }
            runCatching { track.stop() }
            runCatching { track.release() }
        }
    }

    private fun buildSamples(cue: FeedbackCue, volume: Float): ShortArray {
        val gapFrames = (cue.gapMillis * SAMPLE_RATE / 1_000.0).roundToInt()
        val toneFrames = cue.tones.map {
            (it.durationMillis * SAMPLE_RATE / 1_000.0).roundToInt()
        }
        val totalFrames = toneFrames.sum() + max(cue.tones.size - 1, 0) * gapFrames
        if (totalFrames <= 0) return ShortArray(0)

        val samples = ShortArray(totalFrames)
        var offset = 0
        val scaledVolume = (volume * cue.volumeScale).coerceIn(0f, 1f)
        cue.tones.forEachIndexed { index, tone ->
            val frameCount = toneFrames[index]
            val edgeFrames = min(
                (ENVELOPE_MILLIS * SAMPLE_RATE / 1_000.0).roundToInt(),
                max(frameCount / 2, 1),
            )
            for (frame in 0 until frameCount) {
                val attack = min(frame.toFloat() / edgeFrames, 1f)
                val release = min((frameCount - frame - 1).toFloat() / edgeFrames, 1f)
                val envelope = max(min(attack, release), 0f)
                val phase = 2.0 * PI * tone.frequencyHz * frame / SAMPLE_RATE
                var sample = sin(phase)
                if (tone.piercing) {
                    sample += sin(phase * 3.0) / 3.0 + sin(phase * 5.0) / 5.0
                    sample *= 0.85
                }
                val scaled = (sample * tone.amplitude * envelope * scaledVolume)
                    .coerceIn(-1.0, 1.0)
                samples[offset + frame] = (scaled * Short.MAX_VALUE).roundToInt().toShort()
            }
            offset += frameCount
            if (index < cue.tones.lastIndex) offset += gapFrames
        }
        return samples
    }

    private fun playHaptic(haptic: FeedbackHaptic) {
        val vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            context.getSystemService(VibratorManager::class.java)?.defaultVibrator
        } else {
            @Suppress("DEPRECATION")
            context.getSystemService(Vibrator::class.java)
        } ?: return
        if (!vibrator.hasVibrator() || !isHapticFeedbackEnabled()) return

        val effect = when (haptic) {
            FeedbackHaptic.MEDIUM -> VibrationEffect.createOneShot(
                MEDIUM_HAPTIC_MILLIS,
                VibrationEffect.DEFAULT_AMPLITUDE,
            )
            FeedbackHaptic.WARNING -> VibrationEffect.createWaveform(
                longArrayOf(0, WARNING_HAPTIC_MILLIS, WARNING_GAP_MILLIS, WARNING_HAPTIC_MILLIS),
                -1,
            )
            FeedbackHaptic.SUCCESS -> VibrationEffect.createOneShot(
                SUCCESS_HAPTIC_MILLIS,
                VibrationEffect.DEFAULT_AMPLITUDE,
            )
            FeedbackHaptic.ERROR -> VibrationEffect.createWaveform(
                longArrayOf(0, ERROR_HAPTIC_MILLIS, ERROR_GAP_MILLIS, ERROR_HAPTIC_MILLIS),
                -1,
            )
        }
        runCatching { vibrator.vibrate(effect) }
    }

    /** Android's vibrator service also applies the device's global haptic policy. */
    @Suppress("DEPRECATION") // HAPTIC_FEEDBACK_ENABLED is the system user setting on API 31+.
    private fun isHapticFeedbackEnabled(): Boolean = runCatching {
        Settings.System.getInt(
            context.contentResolver,
            Settings.System.HAPTIC_FEEDBACK_ENABLED,
            1,
        ) != 0
    }.getOrDefault(true)

    private fun stopCurrentAudio() {
        val track: AudioTrack?
        val mediaPlayback: MediaPlayback?
        synchronized(this) {
            track = currentTrack
            currentTrack = null
            mediaPlayback = currentMediaPlayback
            currentMediaPlayback = null
        }
        track?.let {
            runCatching { it.pause() }
            runCatching { it.flush() }
            runCatching { it.stop() }
            runCatching { it.release() }
        }
        mediaPlayback?.let(::releaseMediaPlayback)
    }

    /** Idempotent release shared by completion, error, deadline, and new cues. */
    private fun releaseMediaPlayback(playback: MediaPlayback) {
        if (!playback.released.compareAndSet(false, true)) return
        synchronized(this) {
            if (currentMediaPlayback === playback) currentMediaPlayback = null
        }
        playback.deadline?.cancel(false)
        runCatching { playback.player.stop() }
        runCatching { playback.player.release() }
    }

    private companion object {
        const val SAMPLE_RATE = 44_100
        const val ENVELOPE_MILLIS = 6
        const val AUDIO_POLL_MILLIS = 10L
        const val AUDIO_COMPLETION_MARGIN_MILLIS = 250L
        const val MEDIUM_HAPTIC_MILLIS = 45L
        const val WARNING_HAPTIC_MILLIS = 90L
        const val WARNING_GAP_MILLIS = 60L
        const val SUCCESS_HAPTIC_MILLIS = 55L
        const val ERROR_HAPTIC_MILLIS = 90L
        const val ERROR_GAP_MILLIS = 50L
    }
}
